### PR TITLE
Log each condition sent to Z3

### LINKED ISSUE #1361: Log each condition sent to Z3
See `decidePredicate`.

---

- [ ] Log entry for each condition sent to Z3.
- [ ] Insert a unique identifier into the log and the solver transcript. (#1267)

### PR DESCRIPTION
**Pull request by andreiburdusa**
---
Fixes #1361 
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Step/SMT/Evaluator.hs [hunk 1]
```diff
@@ -132,6 +132,10 @@ decidePredicate
     -> simplifier (Maybe Bool)
 decidePredicate korePredicate =
     SMT.withSolver $ runMaybeT $ do
+        (logInfo . Text.pack . show . Pretty.vsep)
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/SMT/Evaluator.hs]: Should we add a new data type and some logging functions similar to the ones in `Kore.Logger.(Debug.*|Warn.*)`, perhaps in `DebugSolver`?

File: kore/src/Kore/Logger/SMTCondition.hs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    ( Predicate
+    )
+import Kore.Internal.TermLike
+import Data.Text.Prettyprint.Doc
+    ( Pretty (..)
+    )
+import Data.Typeable
+import Kore.Unparser
+
+newtype SMTCondition variable =
```
Comment (ttuegel) [inline kore/src/Kore/Logger/SMTCondition.hs]: Please name this `InfoEvaluateCondition`. It must not have any type parameters. The type of the condition should be `Predicate Variable`; this should prevent all the extra constraints.

File: kore/src/Kore/Logger/SMTCondition.hs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    => Entry (SMTCondition variable)
+  where
+    entrySeverity _ = Info
+
+logSMTCondition
+    :: MonadLog log
+    => InternalVariable variable
+    => Typeable variable
+    => Predicate variable -> log ()
+logSMTCondition =
```
Comment (ttuegel) [inline kore/src/Kore/Logger/SMTCondition.hs]: As above, please name this `infoEvaluateCondition`. Here we can allow the `variable` type to be a type variable, but we must externalize the variables; see [here](https://github.com/kframework/kore/blob/master/kore/src/Kore/Internal/TermLike/TermLike.hs#L287) for an example. You will probably need to define a helper in `Kore.Internal.Predicate` that wraps `Kore.Internal.TermLike.externalizeFreshVariables`.

File: kore/src/Kore/Internal/Predicate.hs [hunk 4]
```diff
@@ -765,3 +769,15 @@ substitute
     -> Predicate variable
 substitute subst (GenericPredicate termLike) =
     GenericPredicate (TermLike.substitute subst termLike)
+
```
Comment (ttuegel) [inline kore/src/Kore/Internal/Predicate.hs]:
```suggestion
```

File: kore/src/Kore/Internal/Predicate.hs [hunk 5]
```diff
@@ -765,3 +769,15 @@ substitute
...
+{- | Externalize the variables
+
+-}
+freshVariable
+    :: SortedVariable variable
+    => Predicate variable
+    -> Predicate Variable
+freshVariable predicate =
+    externalizeFreshVariables . (TermLike.mapVariables toVariable)
```
Comment (ttuegel) [inline kore/src/Kore/Internal/Predicate.hs]:
```suggestion
    externalizeFreshVariables . TermLike.mapVariables toVariable
```

**Pull request merged**